proposals/scalable-robust-operator.md

```diff
@@ -0,0 +1,289 @@
+      (scheduling or starting). But Kubernetes SIG Apps is
+      [open to add a status field for running Pods](https://kubernetes.slack.com/archives/C18NZM5K9/p1619549430067400).
+    - Caveat 2: Horovod supports elastic workers, but the Kubernetes Job
+      doesn’t support changes to the completions field. This can be supported
+      starting from 1.23. In the meantime, we can replicate the behavior by
+      creating a new Job and doing Pod adoption.
+  - For Intel MPI, we also need a headless Service to front the launcher,
+    because workers communicate back to the launcher using its hostname.
+- **Revert the use of the Job API for the launcher.**
+  - The Job controller handles retries when the launcher or any of the workers fail.
+  - Caveat 1 also applies: The Job controller doesn’t report if the Pod is running.
+    We can continue watching Pods in the meantime.
+- With the above changes, **the following objects can be removed**:
+  - The ServiceAccount+Role+RoleBinding for the launcher.
+  - The `kubectl-delivery` init container in the launcher, as there is no need
+    to obtain IPs, speeding up startup time.
+    
+## Alternatives Considered
+
+TBD from discussions
+
+## Appendix: Prototype objects
+
+It uses a StatefulSet in place of Indexed Jobs, as they are still an alpha
+feature in Kubernetes.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: mpi-config
+data:
+  hostfile: |
+    mpi-workers-0.mpi-workers slots=3
+    mpi-workers-1.mpi-workers slots=3
+    mpi-workers-2.mpi-workers slots=3
+```
+
+```yaml
+apiVersion: v1
+kind: Secret
+type: kubernetes.io/ssh-auth
+data:
+  ssh-privatekey: PRIVATE_KEY
+  ssh-publickey: PUBLIC_KEY
+```
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: mpi-launcher
+spec:
+  template:
+    spec:
+      restartPolicy: OnFailure
+      containers:
+      - name: driver
+        image: '<USER_IMAGE>'
+        args:
+        - 'mpirun'
+        - '-np'
+        - '9'
+        - '<USER_EXECUTABLE>'
+        env:
+        - name: 'OMPI_MCA_orte_keep_fqdn_hostnames'
+          value: 'true'
+        - name: 'OMPI_MCA_orte_default_hostfile'
+          value: '/home/mpiuser/config/hostfile'
+        volumeMounts:
+        - name: ssh-auth
+          mountPath: /mnt/ssh
+          readOnly: true
+        - name: mpi-config
+          mountPath: /home/mpiuser/config
+          readOnly: true
+      volumes:
+      - name: mpi-config
+        configMap:
+          name: mpi-config
+      - name: ssh-auth
+        secret:
+          secretName: ssh-auth
+          items:
+          - key: ssh-privatekey
+            path: id_rsa
+          - key: ssh-publickey
+            path: id_rsa.pub
+          - key: ssh-publickey
+            path: authorized_keys
+```
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: mpi-workers
+spec:
+  clusterIP: None
+  selector:
+    app: mpi-workers
+```
+
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
```

Comment:
I am wondering how to support elastic mode with statefulset? We may add/remove workers in the fly.

Reply:
You just change `.spec.replicas`.